### PR TITLE
WIP: Add use-vc option to container resolv.conf to force DNS TCP lookups

### PR DESCRIPTION
**Pull request by girishkalele**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28913)

### CODE REVIEW

**Comment by girishkalele** [comment]
@thockin @quinton-hoole 

I added Quinton's suggestion to force DNS TCP lookups.

The code that adds stuff to /etc/resolv.conf for the pods needs a little code factoring to allow adding/removing options on the options: line. Currently the only option is the ndots.

Should the use-vc option be based on some sort of global cluster config somewhere ?

**Comment by quinton-hoole** [comment]
@girishkalele I'd suggest being a little careful with this, and definitely not enabling TCP DNS lookups across the cluster by default.  I'd guess that it will cause higher latency lookups (due to the TCP connection setup delays), and possibly other undesirable side-effects.  Until we can confirm that it's strictly better, I agree, put it in optional config somewhere, and leave the default as UDP.  I guess ideally it should be over-ridable per pod, with cluster-wide default if not specified for a given pod.

File: pkg/kubelet/dockertools/docker_manager.go [hunk 1]
```diff
@@ -79,7 +79,7 @@ const (
 	// ndots specifies the minimum number of dots that a domain name must contain for the resolver to consider it as FQDN (fully-qualified)
 	// we want to able to consider SRV lookup names like _dns._udp.kube-dns.default.svc to be considered relative.
 	// hence, setting ndots to be 5.
-	ndotsDNSOption = "options ndots:5\n"
+	ndotsDNSOption = "options use-vc ndots:5\n"
```
Comment (quinton-hoole) [inline pkg/kubelet/dockertools/docker_manager.go]: Drive-by comment: I don't think that you want 'use-vc' as part of ndotsDNSoption, in the general case, as the two are completely unrelated.

Comment (quinton-hoole) [inline pkg/kubelet/dockertools/docker_manager.go]: I see you already mention this in your PR comments - apologies for the duplication :-)

**Comment by thockin** [comment]
1) I don't think we can propose to make this a default setting without significant trial first.  Have you run dnsbench against this?

2) Yes to some simple refactoring and better comments

3) To make this really configurable, we'd have to get all the way to ConfigMaps for cluster config.  Inztead, you could do an annotation like 'pod.alpha.kubernetes.io/tcp-dns' to trigger kubelet to add this option on a per-pod basis.  We can document it and ask people who are experiencing this problem to try it.  @fabioy @roberthbailey this "alpha" annotation would be one we want to enable for GKE.  Maybe we need a designation that is neither alpha nor beta but instead "needs user trials" ?

4) We should get the users who are reporting problems to try a manual form of this and see if a) it works for them and b) they don't fall over.

**Comment by roberthbailey** [comment]
Why not make it a beta annotation? Do we have concerns keeping compatibility across a couple of versions?

**Comment by girishkalele** [comment]
I will work on a) factoring out the code and 2) making kubelet use the annotation to inject the option into the /etc/resolv.conf.

@roberthbailey IIUC, @thockin wants to setup the alpha annotation as a temporary API till we come up with a more formal and detailed DNS specification for pods.

**Comment by roberthbailey** [comment]
I understood why it would be an annotation, I'm just wondering why it should (or shouldn't) be an alpha vs. a beta annotation.

**Comment by girishkalele** [comment]
alpha vs beta : What are the differences between the two ?
Does a beta annotation mean that we will support it and alpha means we are free to change it ?

**Comment by roberthbailey** [comment]
alpha means it can change at any time. beta means that it won't change behavior unexpectedly. In practice this means that we'd deprecate it (but leave it in) for a release before removing it.

**Comment by thockin** [comment]
I don't mind an alpha annotation, but it would be nice if we had someone,
ANYONE, who said "this works for my problem".
